vncviewer/touch.cxx

```diff
@@ -0,0 +1,627 @@
+      mask = Button2Mask;
+      break;
+    case GH_RIGHTBTN:
+      mask = Button3Mask;
+      break;
+    //case GH_ZOOM:
+    //  mask = ControlMask;
+    //  break;
+    default:
+      mask = 0;
+  }
+
+  fakeEvent.xbutton.state |= mask;
+
+  fl_handle(fakeEvent);
+}
+
+void FLTKGestureHandler::fakeButtonEvent(bool press, int button,
+                                         const XIDeviceEvent* origEvent)
+{
+  XEvent fakeEvent;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = press ? ButtonPress : ButtonRelease;
+  fakeEvent.xbutton.button = button;
+  prepXEvent(&fakeEvent, origEvent);
+  //setTouchEventState(&fakeEvent);
+
+  vlog.info("CORE xbutton.state = %i", fakeEvent.xbutton.state);
+
+  fl_handle(fakeEvent);
+}
+
+void FLTKGestureHandler::fakeButtonEvent(bool press, int button,
+                                         const GHEvent origEvent)
+{
+  XEvent fakeEvent;
+  int mask;
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = press ? ButtonPress : ButtonRelease;
+  fakeEvent.xbutton.button = button;
+  prepXEvent(&fakeEvent, origEvent);
+
+  switch (button) {
+    case Button1:
+      mask = Button1Mask;
+      break;
+    case Button2:
+      mask = Button2Mask;
+      break;
+    case Button3:
+      mask = Button3Mask;
+      break;
+    case Button4:
+      mask = Button4Mask;
+      break;
+    case Button5:
+      mask = Button5Mask;
+      break;
+    case Button6:
+      mask = Button6Mask;
+      break;
+    case Button7:
+      mask = Button7Mask;
+      break;
+    default:
+      mask = 0;
+  }
+
+  if (press)
+    fakeEvent.xbutton.state |= mask;
+  else
+    fakeEvent.xbutton.state &= ~mask;
+
+  fl_handle(fakeEvent);
+}
+
+void FLTKGestureHandler::fakeKeyEvent(bool press, int keysym,
+                                      const GHEvent origEvent)
+{
+  XEvent fakeEvent;
+
+  Window root, child;
+  int x_root, y_root;
+  XkbStateRec state;
+
+  root = XDefaultRootWindow(fl_display);
+  XTranslateCoordinates(fl_display, wnd, root,
+                        origEvent.event_x,
+                        origEvent.event_y,
+                        &x_root, &y_root, &child);
+  XkbGetState(fl_display, XkbUseCoreKbd, &state);
+
+  KeyCode kc = XKeysymToKeycode(fl_display, keysym);
+
+  memset(&fakeEvent, 0, sizeof(XEvent));
+
+  fakeEvent.type = press ? KeyPress : KeyRelease;
+  fakeEvent.xkey.type = press ? KeyPress : KeyRelease;
+  fakeEvent.xkey.keycode = kc;
+  fakeEvent.xkey.serial = XLastKnownRequestProcessed(fl_display);
+  fakeEvent.xkey.display = fl_display;
+  fakeEvent.xkey.window = wnd;
+  fakeEvent.xkey.root = root;
+  fakeEvent.xkey.subwindow = None;
+  fakeEvent.xkey.time = fl_event_time;
+  fakeEvent.xkey.x = origEvent.event_x;
+  fakeEvent.xkey.y = origEvent.event_y;
+  fakeEvent.xkey.x_root = x_root;
+  fakeEvent.xkey.y_root = y_root;
+  //fakeEvent.xkey.state = state.mods;
+  //fakeEvent.xkey.state |= ((state.ptr_buttons >> 1) & 0x1f) << 8; // FIXME: Check this!
```

Comment:
This should be fixed so we don't have surprises in the future.

To simplify things I would suggest going away from a generic key faking and focusing on just faking Ctrl. You can take a lot of shortcuts in that case.

Reply:
I'm not sure what the fix is here, since I don't understand that code. But according to `xev` testing we should only be setting the mask for key release events anyway.

I struggle with these masks and states to be honest, I don't quite understand how they're supposed to work in our situation. Ideally I'd like to get rid of `fakeKeyEvent()` completely - since as you say, we only need it for Ctrl - and just set the Ctrl bit of the modifier mask for those events which need it. But this doesn't work, you seem to need the keypress event. Further, the modifiers don't seem to matter at all, as evidenced by the fact that it's completely commented out here and everything still works. So I'm really unsure about what the "fix" is here.

Reply:
The state field contains all the modifiers (e.g. Shift, Ctrl) and mouse buttons that are currently pressed. FLTK might not look at it right now (or we haven't tested it fully), but it might in the future. I dread trying to debug the weird effects that might have, that's why I insist we should attempt to do this properly now.